File: python/mirascope/llm/tools/web_search_tool.py
"""Web search tool for provider-native web search capabilities."""

from dataclasses import dataclass, field

from .provider_tools import ProviderTool


@dataclass(frozen=True)
class WebSearchTool(ProviderTool):
    """Web search tool that allows the model to search the internet.

    This is a provider tool - the search is executed server-side by the provider,
    not by your code. The model decides when to search based on the prompt,
    and the provider returns search results with citations.

    Supported providers include Anthropic, Google, and OpenAI (when using the Responses API).

    Example:
        ```python
        from mirascope import llm

        @llm.call("anthropic/claude-sonnet-4-5", tools=[llm.WebSearchTool()])
        def search_web(query: str) -> str:
            return f"Search the web for: {query}"

        response = search_web("Who won the 2024 Super Bowl?")
        print(response.text())  # Response includes citations from web search
        ```
    """

    name: str = field(default="web_search", init=False)
    """The tool name. Always "web_search" for this tool type."""

File: typescript/src/llm/tools/web-search-tool.ts
/**
 * Web search tool for provider-native web search capabilities.
 */

import { ProviderTool } from "./provider-tool";

/**
 * Web search tool that allows the model to search the internet.
 *
 * This is a provider tool - the search is executed server-side by the provider,
 * not by your code. The model decides when to search based on the prompt,
 * and the provider returns search results with citations.
 *
 * Supported providers include Anthropic, Google, and OpenAI (when using the Responses API).
 *
 * @example
 * ```typescript
 * import { defineCall, WebSearchTool } from '@anthropic-ai/mirascope';
 *
 * const searchWeb = defineCall({
 *   model: 'anthropic/claude-sonnet-4-5',
 *   tools: [new WebSearchTool()],
 * }, () => 'Search the web for: Who won the 2024 Super Bowl?');
 *
 * const response = await searchWeb();
 * console.log(response.text()); // Response includes citations from web search
 * ```
 */
export class WebSearchTool extends ProviderTool {
  constructor() {
    super("web_search");
  }
}

/**
 * Type guard to check if a value is a WebSearchTool.
 */
export function isWebSearchTool(value: unknown): value is WebSearchTool {
  return value instanceof WebSearchTool;
}
